L2-Theorems.agda

```diff
@@ -69,3 +76,131 @@ Substitution {Σ} deriv = ⊢-induction-simple case deriv where
     case (app h₁ h₂)                       = app h₁ h₂
     case (letval h₁ h₂)                    = letval h₁ h₂
     case (letrecfn h₁ h₂)                  = letrecfn h₁ h₂
+
+dom⊆ : StoreEnv → Store → Set
+dom⊆ Σ s = ∀ {ℓ} → Σ ℓ ≡ just intref → ∃[ z ] ((s !! ℓ) ≡ just z)
+
+refAssignSame : ∀ {n} s ℓ → (s ⨄ (ℓ ↦ n)) !! ℓ ≡ just n
+refAssignSame    []    zero  = refl
+refAssignSame (_ ∷ _)  zero  = refl
+refAssignSame    []   (suc ℓ) = refAssignSame [] ℓ
+refAssignSame (_ ∷ s) (suc ℓ) = refAssignSame s ℓ
+
+refAssignDiff : ∀ {ℓ ℓ' n} → (s : Store) → ¬ (ℓ ≡ ℓ') → ((s ⨄ (ℓ ↦ n)) !! ℓ') ≡ s !! ℓ'
+refAssignDiff {zero} {zero} s ¬p = ⊥-elim (¬p refl)
+refAssignDiff {zero} {suc ℓ'} [] _ = refl
+refAssignDiff {zero} {suc ℓ'} (_ ∷ []) _ = refl
+refAssignDiff {zero} {suc ℓ'} (_ ∷ _ ∷ _) _ = refl
+refAssignDiff {suc ℓ} {zero} [] _ = refl
+refAssignDiff {suc ℓ} {zero} (_ ∷ _) _ = refl
+refAssignDiff {suc ℓ} {suc ℓ'} [] ¬p = refAssignDiff [] (contraposition (cong suc) ¬p)
+refAssignDiff {suc ℓ} {suc ℓ'} (_ ∷ s) ¬p = refAssignDiff s (contraposition (cong suc) ¬p)
+
+dom⊆-extend : ∀ {ℓ Σ n} s → dom⊆ Σ s → dom⊆ Σ (s ⨄ (ℓ ↦ n))
+dom⊆-extend [] d⊆ eq with ⟨ _ , () ⟩ ← d⊆ eq
+dom⊆-extend {ℓ} (x ∷ s) d⊆ {ℓ'} eq with ℓ ≟ ℓ' | d⊆ eq
+... | yes refl | _  = ⟨ _ , refAssignSame (x ∷ s) ℓ ⟩
+dom⊆-extend s d⊆ _ | no ¬p | ⟨ fst , eq' ⟩ = ⟨ fst , trans (refAssignDiff s ¬p) eq' ⟩
+
+data val-or-step (s : Store) (e : Expression) : Set where
+  val : Value e →  val-or-step s e
+  step : ∀ {e' s'} → ⟨ e , s ⟩ ⟶ ⟨ e' , s' ⟩ → val-or-step s e
+
+-- Theorem 18: Progress
+Progress : ∀ {Σ Γ e T s} → Σ ⨾ Γ ⊢ e ∶ T → dom⊆ Σ s → Γ ≡ • → val-or-step s e
```

Comment:
I think the proof might be easier without a separate equality proof `Γ ≡ •`, i.e. if you move the assumption that the environment is empty into the derivation:

```agda
Progress : ∀ {Σ e T s} → Σ ⨾ • ⊢ e ∶ T → dom⊆ Σ s → val-or-step s e
```